### PR TITLE
V5 compatibility and features

### LINKED ISSUE #18: Plugin hooks deprecation
Just in case you hadn't seen this yet...

It looks like the psalm plugin hooks have been deprecated and could be removed as early as v5 (https://github.com/vimeo/psalm/issues/4700#issuecomment-855350737).

I suppose plugins are meant to use eventhandlers instead.

I didn't see anything anywhere about a timeline for v5 release.

### PR DESCRIPTION
**Pull request by kkmuffme**
@mcaskill finally I had some time to PR it; my fork was quite different already and I had to remove a couple things that are internal, so it took a bit more time.

It's been almost 2 years since I wrote some of the code, and I'm much more familiar with psalm's code base now than I was then, so I already see a variety of things that can be improved upon eventually.

If you could take a quick look and run this branch for a few days in your codebase, to identify any issues/points you want to change upon, that would be great.

Otherwise, I'd say we merge this, and then add/merge the few changes of https://github.com/psalm/psalm-plugin-wordpress/pull/42/files, which aren't part of this PR already, to it (if you don't want to do that, I'm fine to do that chore)

--

This PR adds compatibility with psalm v5, new phpparser version (afair 4.16 or something), support for dynamic hook names, some support for cron hooks, errors for deprecated hooks, handling for hooks without docs, and tons of other stuff.

Fix https://github.com/psalm/psalm-plugin-wordpress/issues/18

### CODE REVIEW

File: composer.json
Comment (mcaskill) [inline composer.json:1]: [psalm/plugin-phpunit](https://github.com/psalm/psalm-plugin-phpunit) should require [`^0.18.4`](https://github.com/psalm/psalm-plugin-phpunit/releases/tag/0.18.4) which allows Psalm 5.

File: Plugin.php
Comment (mcaskill) [inline Plugin.php:1]: There's a handful of occurrences of inconsistent coding style.

```diff
- });
+ } );
```

File: Plugin.php
Comment (mcaskill) [inline Plugin.php]: For consistency, there should be a comment for `requireAllParams`, similar to the ones for the other options.
Furthermore, the `requireAllParams` option should be documented in the `README.md`.

File: Plugin.php
Comment (mcaskill) [inline Plugin.php:1]: There's a few changes on the `master` branch that you are going to need to rebase/include:

* c4dcb85 — Add `hooks.json` file as candidate for custom hooks.
* bd97269 — Add `stubs/globals.php` to stub files (#44)

These changes should satisfy the branch conflicts that GitHub has raised.

**Review by mcaskill — CHANGES_REQUESTED** [review]
I've performed a cursory review of the changes. Looks good. A lot to absorb. I've perform a deeper dive as soon as I can.

File: Plugin.php
Comment (kkmuffme) [reply Plugin.php:1]: Master branch is broken compeltely tbh - https://github.com/psalm/psalm-plugin-wordpress/commit/b6ceeefce947c0d60170aa7f2496334fbc6761b0 the @var you added was wrong and it's not loading the hooks at all, bc this is a SimpleXMLElement, not a string.

If ixed this in this branch now too.

Will rebase it anyway though

Comment (kkmuffme) [reply Plugin.php:1]: Unable to find where this is

File: Plugin.php
Comment (kkmuffme) [reply Plugin.php]: Thanks, done!

File: Plugin.php
Comment (mcaskill) [reply Plugin.php:1]: Let's leave this thread open. I'll do a pass of edits for coding style.

Comment (paulshryock) [reply Plugin.php:1]: Is this not something that `composer cs-fix` would fix?

Comment (mcaskill) [reply Plugin.php:1]: Most likely. Once we know everything else is working.

Comment (kkmuffme) [reply Plugin.php:1]: I think we should add a pipeline for these cs fixes for github later, just like psalm has.
And maybe also change the cs ruleset to psalm's as that would make more sense, bc now it's some weird in between of WP and psalm.

At a later point in master branch though, not part of this PR.

Comment (kkmuffme) [reply Plugin.php:1]: @mcaskill @paulshryock any further questions, otherwise I think we can merge this and release a new version

For the CI stuff I created an issue https://github.com/psalm/psalm-plugin-wordpress/issues/50 so this can be tackled after the merge
I guess it's straightforward and shouldn't take long, so we can add that and finally create a release in the next 2 weeks before the end of the year.

Comment (mcaskill) [reply Plugin.php:1]: Good idea with the separate issue.

Unfortunately, after all, I won't be able to test this PR until maybe January.
You have my approval to merge this PR. Should be good.

**Review by mcaskill — APPROVED** [review]

File: Plugin.php
Comment (joehoyle) [inline Plugin.php:944]: This seems like a very bizarre error to me. Is the expectation that every have to do `add_action( 'init', 'my_plugin_init', 10, 0 )`?

Comment (kkmuffme) [reply Plugin.php:944]: Yes. This isn't a 100% necessity, but so are most issues on psalm level 0/Mixed issues, I guess this could be made to only report on the strictest level

Comment (kkmuffme) [reply Plugin.php:944]: I remembered why: depending on the callback used, this could result in a fatal error:
ArgumentCountError: ... expects at most 0 arguments, 1 given

and this error is something that is otherwise reported by psalm too (TooManyArguments or something)

Comment (joehoyle) [reply Plugin.php:944]: Ah I see-- do we inspect the hook though to see if it _will_ actually pass an arg, rather than just relying on the default? E.g. I think "init" doesn't pass any args

Comment (kkmuffme) [reply Plugin.php:944]: >do we inspect the hook though to see if it will actually pass an arg, rather than just relying on the default?

Afair no, and the reason for that was that I got burnt with that before: various plugins (including WooCommerce at times) sometimes call hooks with a different number of args, which again resulted in a (rare) fatal error.
Besides that, it's risky with dynamic hook names to get a false negative.
Basically, setting the 0 parameter is just ensuring you won't randomly get a fatal

**Pull request merged**
